File: packages/python/goldenmatch/goldenmatch/a2a/skills.py
"""Skill dispatch for the A2A protocol server.

Routes incoming skill requests to the appropriate AgentSession methods
or GoldenMatch API functions.
"""

from __future__ import annotations

from typing import Any

import yaml

from goldenmatch.core.agent import (
    AgentSession,
    _decision_to_config,
    profile_for_agent,
    select_strategy,
)


def dispatch_skill(skill_id: str, params: dict) -> dict:
    """Dispatch an A2A skill request to the appropriate handler.

    Parameters
    ----------
    skill_id : str
        One of: analyze_data, configure, deduplicate, match, explain,
        review, compare_strategies, pprl, quality, transform.
    params : dict
        Skill-specific parameters.

    Returns
    -------
    dict
        Result payload.

    Raises
    ------
    ValueError
        If *skill_id* is not recognised.
    """
    session = AgentSession()

    if skill_id == "analyze_data":
        return session.analyze(params["file_path"])

    if skill_id == "autoconfig":
        # v1.7-v1.12: AutoConfigController via AgentSession. Returns committed
        # config + telemetry blob (stop_reason, health, decisions, NE fields).
        excl = params.get("exclude_columns") or None
        _excl_token = None
        if excl:
            from goldenmatch.core.autoconfig import _RUNTIME_EXCLUDE_COLUMNS
            _excl_token = _RUNTIME_EXCLUDE_COLUMNS.set(list(excl))
        try:
            return session.autoconfigure(params["file_path"])
        finally:
            if _excl_token is not None:
                from goldenmatch.core.autoconfig import _RUNTIME_EXCLUDE_COLUMNS
                _RUNTIME_EXCLUDE_COLUMNS.reset(_excl_token)

    if skill_id == "controller_telemetry":
        # Stateless A2A dispatch (one AgentSession per request) means we
        # can't recall telemetry from a prior request. Mirrors MCP's note —
        # callers should use the inline telemetry returned by autoconfig /
        # deduplicate / match instead.
        return {
            "available": False,
            "note": (
                "controller_telemetry is per-session and A2A dispatch is "
                "stateless. Telemetry is returned inline by autoconfig, "
                "deduplicate, and match skills."
            ),
        }

    if skill_id == "deduplicate":
        excl = params.get("exclude_columns") or None
        _excl_token = None
        if excl:
            from goldenmatch.core.autoconfig import _RUNTIME_EXCLUDE_COLUMNS
            _excl_token = _RUNTIME_EXCLUDE_COLUMNS.set(list(excl))
        try:
            result = session.deduplicate(
                params["file_path"],
                config=params.get("config"),
            )
        finally:
            if _excl_token is not None:
                from goldenmatch.core.autoconfig import _RUNTIME_EXCLUDE_COLUMNS
                _RUNTIME_EXCLUDE_COLUMNS.reset(_excl_token)
        serialised = _serialise_result(result)
        # Plumb telemetry onto the wire result so callers see stop_reason /
        # decisions / NE alongside the dedupe output, not as a separate call.
        serialised["telemetry"] = session.last_telemetry or {"available": False, "source": None}
        return serialised

    if skill_id == "match":
        excl = params.get("exclude_columns") or None
        _excl_token = None
        if excl:
            from goldenmatch.core.autoconfig import _RUNTIME_EXCLUDE_COLUMNS
            _excl_token = _RUNTIME_EXCLUDE_COLUMNS.set(list(excl))
        try:
            result = session.match_sources(
                params["file_a"],
                params["file_b"],
                config=params.get("config"),
            )
        finally:
            if _excl_token is not None:
                from goldenmatch.core.autoconfig import _RUNTIME_EXCLUDE_COLUMNS
                _RUNTIME_EXCLUDE_COLUMNS.reset(_excl_token)
        serialised = _serialise_result(result)
        serialised["telemetry"] = session.last_telemetry or {"available": False, "source": None}
        return serialised

    if skill_id == "compare_strategies":
        return session.compare_strategies(
            params["file_path"],
            ground_truth=params.get("ground_truth"),
        )

    if skill_id == "explain":
        import polars as pl

        from goldenmatch import explain_pair_df

        record_a = pl.DataFrame([params["record_a"]])
        record_b = pl.DataFrame([params["record_b"]])
        mk_cfg = params["matchkey"]
        explanation = explain_pair_df(record_a, record_b, mk_cfg)
        return {"explanation": explanation}

    if skill_id == "review":
        from goldenmatch.core.review_queue import ReviewQueue

        queue = ReviewQueue(backend="memory")
        return {"pending": queue.list_pending()}

    if skill_id == "configure":
        import polars as pl

        _analysis = session.analyze(params["file_path"])
        decision = select_strategy(
            profile_for_agent(
                pl.read_csv(params["file_path"], encoding="utf8-lossy", ignore_errors=True)
            )
        )
        cfg = _decision_to_config(decision)
        return {"config_yaml": yaml.dump(cfg.model_dump(), default_flow_style=False)}

    if skill_id == "pprl":
        from goldenmatch import pprl_link

        result = pprl_link(
            params["file_a"],
            params["file_b"],
            fields=params.get("fields", []),
        )
        return _serialise_result({"result": result})

    if skill_id == "quality":
        import polars as pl

        from goldenmatch.config.schemas import QualityConfig
        from goldenmatch.core.quality import _goldencheck_available, run_quality_check

        if not _goldencheck_available():
            return {"error": "goldencheck not installed. pip install goldenmatch[quality]"}

        file_path = params.get("file_path")
        if not file_path:
            return {"error": "Missing required parameter: file_path"}

        try:
            df = pl.read_csv(file_path, encoding="utf8-lossy", ignore_errors=True)
        except FileNotFoundError:
            return {"error": f"File not found: {file_path}"}
        except Exception as exc:
            return {"error": f"Could not read CSV '{file_path}': {exc}"}

        fix_mode = params.get("fix_mode", "safe")
        domain = params.get("domain")
        qc = QualityConfig(mode="silent", fix_mode=fix_mode, domain=domain)
        fixed_df, fixes = run_quality_check(df, qc)

        output_path = params.get("output_path")
        write_error = None
        if output_path:
            try:
                fixed_df.write_csv(output_path)
            except Exception as exc:
                write_error = f"Results computed but failed to write to '{output_path}': {exc}"
                output_path = None

        result = {
            "total_records": fixed_df.height,
            "fixes_applied": len(fixes),
            "fixes": fixes,
            "output_path": output_path,
        }
        if write_error:
            result["write_error"] = write_error
        return result

    if skill_id == "transform":
        import polars as pl

        from goldenmatch.config.schemas import TransformConfig
        from goldenmatch.core.transform import _goldenflow_available, run_transform

        if not _goldenflow_available():
            return {"error": "goldenflow not installed. pip install goldenmatch[transform]"}

        file_path = params.get("file_path")
        if not file_path:
            return {"error": "Missing required parameter: file_path"}

        try:
            df = pl.read_csv(file_path, encoding="utf8-lossy", ignore_errors=True)
        except FileNotFoundError:
            return {"error": f"File not found: {file_path}"}
        except Exception as exc:
            return {"error": f"Could not read CSV '{file_path}': {exc}"}

        tc = TransformConfig(mode="silent")
        transformed_df, fixes = run_transform(df, tc, strict=True)

        output_path = params.get("output_path")
        write_error = None
        if output_path:
            try:
                transformed_df.write_csv(output_path)
            except Exception as exc:
                write_error = f"Results computed but failed to write to '{output_path}': {exc}"
                output_path = None

        result = {
            "total_records": transformed_df.height,
            "transforms_applied": len(fixes),
            "transforms": fixes,
            "output_path": output_path,
        }
        if write_error:
            result["write_error"] = write_error
        return result

    if skill_id in {
        "identity_resolve", "identity_list", "identity_history",
        "identity_conflicts", "identity_merge", "identity_split",
    }:
        from goldenmatch.mcp.identity_tools import _dispatch as _identity_dispatch
        # Reuse MCP dispatch since the contract is identical (JSON in/out).
        return _identity_dispatch(skill_id, params)

    if skill_id == "add_correction":
        # v1.19.x Phase 3 (#437 surface sync). Pair-level OR field-level.
        # Inlined dispatch (rather than delegating to MCP) so this branch is
        # independent of the Phase 1 PR landing order.
        import uuid as _uuid
        from datetime import datetime as _dt

        from goldenmatch.core.memory.store import (
            Correction as _CorrectionDC,
        )
        from goldenmatch.core.memory.store import (
            MemoryStore as _MemoryStore,
        )
        from goldenmatch.core.memory.store import (
            _canon_pair,
        )

        dataset = params.get("dataset")
        if not dataset:
            return {"error": "dataset is required"}
        decision = params.get("decision")
        if decision not in ("approve", "reject", "field_correct"):
            return {"error": f"Invalid decision: {decision!r}"}
        path = params.get("path") or ".goldenmatch/memory.db"

        if decision == "field_correct":
            field_name = params.get("field_name")
            corrected_value = params.get("corrected_value")
            if not field_name:
                return {"error": "field_correct requires field_name"}
            if corrected_value is None:
                return {"error": "field_correct requires corrected_value"}
            cid = params.get("cluster_id", params.get("id_a", 0))
            try:
                cid = int(cid)
            except (TypeError, ValueError) as exc:
                return {"error": f"cluster_id must be an integer: {exc}"}
            correction = _CorrectionDC(
                id=str(_uuid.uuid4()),
                id_a=cid,
                id_b=0,
                decision=decision,
                source="agent",
                trust=0.5,
                field_hash="",
                record_hash="",
                original_score=0.0,
                matchkey_name=params.get("matchkey_name"),
                reason=params.get("reason"),
                dataset=dataset,
                created_at=_dt.now(),
                field_name=field_name,
                original_value=params.get("original_value"),
                corrected_value=corrected_value,
            )
        else:
            try:
                id_a = int(params["id_a"])
                id_b = int(params["id_b"])
            except (KeyError, TypeError, ValueError) as exc:
                return {"error": f"id_a / id_b must be integers: {exc}"}
            ca, cb = _canon_pair(id_a, id_b)
            correction = _CorrectionDC(
                id=str(_uuid.uuid4()),
                id_a=ca,
                id_b=cb,
                decision=decision,
                source="agent",
                trust=0.5,
                field_hash="",
                record_hash="",
                original_score=0.0,
                matchkey_name=params.get("matchkey_name"),
                reason=params.get("reason"),
                dataset=dataset,
                created_at=_dt.now(),
            )

        with _MemoryStore(backend="sqlite", path=path) as store:
            store.add_correction(correction)
        result: dict = {
            "status": "ok",
            "id": correction.id,
            "decision": decision,
            "source": "agent",
            "trust": 0.5,
            "dataset": dataset,
        }
        if decision == "field_correct":
            result["cluster_id"] = correction.id_a
            result["field_name"] = correction.field_name
            result["corrected_value"] = correction.corrected_value
        else:
            result["id_a"] = correction.id_a
            result["id_b"] = correction.id_b
        return result

    raise ValueError(f"Unknown skill: {skill_id}")


def _serialise_result(obj: Any) -> dict:
    """Best-effort serialisation of pipeline results to JSON-safe dict."""
    if isinstance(obj, dict):
        out = {}
        for k, v in obj.items():
            try:
                import polars as pl

                if isinstance(v, pl.DataFrame):
                    out[k] = {"rows": v.height, "columns": v.columns}
                    continue
            except ImportError:
                pass
            if isinstance(v, dict):
                out[k] = _serialise_result(v)
            elif isinstance(v, (str, int, float, bool, type(None))):
                out[k] = v
            elif isinstance(v, list):
                out[k] = str(v)[:500]
            else:
                out[k] = str(v)[:500]
        return out
    return {"value": str(obj)[:500]}

File: packages/typescript/goldenmatch/src/core/agent/skills.ts
/**
 * skills.ts — AGENT_SKILLS registry + dispatchSkill (the shared dispatcher).
 * Edge-safe: no Node.js imports. The I/O seam is dependency injection
 * (`ctx.loadTable`), never a file read here.
 *
 * Wave 1 registers only the AgentSession-backed skills. MCP/A2A wiring, file
 * loaders, and optional-dependency tools (scan_quality, run_transforms, ...)
 * land in Waves 2-4 — the registry is shaped so they slot in later.
 */

import type {
  Row,
  MatchkeyConfig,
  MatchkeyField,
  GoldenMatchConfig,
} from "../types.js";
import { makeMatchkeyConfig, makeMatchkeyField } from "../types.js";
import type { SkillDef, SkillContext, SkillResult, JSONSchema } from "./types.js";
import type { Correction } from "../memory/types.js";
import { profileForAgent, selectStrategy } from "./strategy.js";
import { explainPair } from "../explain.js";
import { gatePairs } from "../review-queue.js";
import { dedupe } from "../api.js";
import { reviewConfig, serializeSuggestions } from "../suggest.js";
import { autoConfigureRows } from "../autoconfig.js";

// ---------------------------------------------------------------------------
// rows-or-path seam
// ---------------------------------------------------------------------------

/**
 * Resolve a table from `args.rows` (inline, edge path) or — only when rows are
 * absent — from `ctx.loadTable(args.file_path)`. Throws a clear error when
 * neither is usable (caught by the dispatcher).
 */
async function resolveTable(
  args: Record<string, unknown>,
  ctx: SkillContext,
  rowsKey = "rows",
  pathKey = "file_path",
): Promise<Row[]> {
  const inline = args[rowsKey];
  if (Array.isArray(inline)) {
    return inline as Row[];
  }
  const path = args[pathKey];
  if (typeof path === "string") {
    return ctx.loadTable(path);
  }
  // No inline rows and no path: still call loadTable so an injected loader can
  // raise the surface-specific "no loader" error (matches the test contract).
  return ctx.loadTable("");
}

/**
 * Coerce a tool argument to a finite integer, or null when it isn't one.
 * Mirrors the `add_correction` MCP tool's `id_a` / `id_b` parsing.
 */
function toInt(raw: unknown): number | null {
  const n = typeof raw === "number" ? raw : parseInt(String(raw), 10);
  return Number.isFinite(n) ? n : null;
}

// ---------------------------------------------------------------------------
// Shared input schemas (reused verbatim by the MCP Tool + A2A card in Wave 2/3)
// ---------------------------------------------------------------------------

const ROWS_OR_PATH_SCHEMA: JSONSchema = {
  type: "object",
  properties: {
    rows: {
      type: "array",
      description: "Inline row objects (edge path). Takes precedence over file_path.",
      items: { type: "object" },
    },
    file_path: {
      type: "string",
      description: "Path to a CSV table, loaded via the injected loader.",
    },
  },
};

const MATCH_SOURCES_SCHEMA: JSONSchema = {
  type: "object",
  properties: {
    rows: { type: "array", items: { type: "object" } },
    file_path: { type: "string" },
    rows_b: { type: "array", items: { type: "object" } },
    file_path_b: { type: "string" },
  },
};

const EXPLAIN_PAIR_SCHEMA: JSONSchema = {
  type: "object",
  properties: {
    record_a: { type: "object", description: "First record (column -> value)." },
    record_b: { type: "object", description: "Second record (column -> value)." },
    fuzzy: {
      type: "object",
      additionalProperties: { type: "number" },
      description: "Map of field -> weight scored with jaro_winkler.",
    },
    exact: {
      type: "array",
      items: { type: "string" },
      description: "Field names compared with the exact scorer.",
    },
  },
  required: ["record_a", "record_b"],
};

const EXPLAIN_CLUSTER_SCHEMA: JSONSchema = {
  type: "object",
  properties: {
    cluster_id: { type: "integer" },
  },
  required: ["cluster_id"],
};

const CONTROLLER_TELEMETRY_SCHEMA: JSONSchema = {
  type: "object",
  properties: {},
};

const APPROVE_REJECT_SCHEMA: JSONSchema = {
  type: "object",
  properties: {
    job_name: { type: "string" },
    id_a: { type: "integer" },
    id_b: { type: "integer" },
    decision: { type: "string", enum: ["approve", "reject"] },
    decided_by: { type: "string" },
    reason: { type: "string" },
  },
  required: ["id_a", "id_b", "decision"],
};

const SCAN_QUALITY_SCHEMA: JSONSchema = {
  type: "object",
  properties: {
    rows: {
      type: "array",
      description: "Inline row objects (edge path). Takes precedence over file_path.",
      items: { type: "object" },
    },
    file_path: {
      type: "string",
      description: "Path to a CSV table, loaded via the injected loader.",
    },
    domain: {
      type: "string",
      description: "Optional domain hint (healthcare, finance, ecommerce).",
    },
  },
};

const FIX_QUALITY_SCHEMA: JSONSchema = {
  type: "object",
  properties: {
    rows: {
      type: "array",
      description: "Inline row objects (edge path). Takes precedence over file_path.",
      items: { type: "object" },
    },
    file_path: {
      type: "string",
      description: "Path to a CSV table, loaded via the injected loader.",
    },
    fix_mode: {
      type: "string",
      enum: ["safe", "moderate"],
      description: "Fix aggressiveness: safe (conservative) or moderate. Default: safe.",
    },
    domain: {
      type: "string",
      description: "Optional domain hint (healthcare, finance, ecommerce).",
    },
  },
};

const REVIEW_CONFIG_SCHEMA: JSONSchema = {
  type: "object",
  properties: {
    rows: {
      type: "array",
      description: "Inline row objects (edge path). Takes precedence over file_path.",
      items: { type: "object" },
    },
    file_path: {
      type: "string",
      description: "Path to a CSV table, loaded via the injected loader.",
    },
    config: {
      type: "object",
      description:
        "Optional GoldenMatchConfig to review. Auto-configured from the data when omitted.",
    },
    verify: {
      type: "boolean",
      description:
        "Re-run the pipeline per candidate and drop health-worsening suggestions (default true).",
    },
  },
};

// ---------------------------------------------------------------------------
// agent_explain_pair: build a MatchkeyConfig from fuzzy/exact args
// ---------------------------------------------------------------------------

/**
 * Build the MatchkeyConfig that `explainPair` requires from the tool's
 * `fuzzy` (`{field: weight}`) and `exact` (`string[]`) args. Mirrors the
 * Python `explain_pair_df(record_a, record_b, fuzzy=, exact=)` shape.
 *
 * When neither is supplied, every shared key of the two records is compared
 * with `jaro_winkler` (a useful default for an ad-hoc pair explanation).
 */
function buildExplainMatchkey(
  rowA: Row,
  rowB: Row,
  fuzzy: unknown,
  exact: unknown,
): MatchkeyConfig {
  const fields: MatchkeyField[] = [];

  if (fuzzy !== null && typeof fuzzy === "object" && !Array.isArray(fuzzy)) {
    for (const [col, weight] of Object.entries(fuzzy as Record<string, unknown>)) {
      const w = typeof weight === "number" ? weight : Number(weight);
      fields.push(
        makeMatchkeyField({
          field: col,
          transforms: ["lowercase", "strip"],
          scorer: "jaro_winkler",
          weight: Number.isFinite(w) ? w : 1.0,
        }),
      );
    }
  }

  if (Array.isArray(exact)) {
    for (const col of exact) {
      fields.push(
        makeMatchkeyField({
          field: String(col),
          transforms: ["lowercase", "strip"],
          scorer: "exact",
        }),
      );
    }
  }

  if (fields.length === 0) {
    const shared = Object.keys(rowA).filter((k) => k in rowB);
    for (const col of shared) {
      fields.push(
        makeMatchkeyField({
          field: col,
          transforms: ["lowercase", "strip"],
          scorer: "jaro_winkler",
        }),
      );
    }
  }

  return makeMatchkeyConfig({
    name: "adhoc",
    type: "weighted",
    fields,
    threshold: 0.85,
  });
}

// ---------------------------------------------------------------------------
// Skill registry
// ---------------------------------------------------------------------------

export const AGENT_SKILLS: readonly SkillDef[] = [
  {
    id: "analyze_data",
    description:
      "Profile a dataset and recommend a matching strategy, with reasoning and alternatives.",
    inputSchema: ROWS_OR_PATH_SCHEMA,
    handler: async (args, ctx): Promise<SkillResult> => {
      const rows = await resolveTable(args, ctx);
      return ctx.session.analyze(rows) as unknown as SkillResult;
    },
  },
  {
    id: "auto_configure",
    description:
      "Run the iterative auto-config controller and return the committed config + telemetry.",
    inputSchema: ROWS_OR_PATH_SCHEMA,
    handler: async (args, ctx): Promise<SkillResult> => {
      const rows = await resolveTable(args, ctx);
      const out = await ctx.session.autoconfigure(rows);
      return { config: out.config, telemetry: out.telemetry };
    },
  },
  {
    id: "agent_deduplicate",
    description:
      "Deduplicate a dataset end-to-end: profile, strategy, gating, and review-queue accumulation.",
    inputSchema: ROWS_OR_PATH_SCHEMA,
    handler: async (args, ctx): Promise<SkillResult> => {
      const rows = await resolveTable(args, ctx);
      const out = await ctx.session.deduplicate(rows);
      return {
        results: out.results,
        reasoning: out.reasoning,
        confidence_distribution: out.confidence_distribution,
        storage: out.storage,
      };
    },
  },
  {
    id: "agent_match_sources",
    description:
      "Match two datasets (target vs reference) with profiling, strategy, and reasoning.",
    inputSchema: MATCH_SOURCES_SCHEMA,
    handler: async (args, ctx): Promise<SkillResult> => {
      const rowsA = await resolveTable(args, ctx, "rows", "file_path");
      const rowsB = await resolveTable(args, ctx, "rows_b", "file_path_b");
      const out = await ctx.session.matchSources(rowsA, rowsB);
      return { results: out.results, reasoning: out.reasoning };
    },
  },
  {
    id: "agent_compare_strategies",
    description:
      "Run multiple candidate strategies on the same dataset and compare proxy metrics.",
    inputSchema: ROWS_OR_PATH_SCHEMA,
    handler: async (args, ctx): Promise<SkillResult> => {
      const rows = await resolveTable(args, ctx);
      const gt = args.ground_truth as
        | readonly (readonly [number, number])[]
        | undefined;
      const out =
        gt !== undefined
          ? await ctx.session.compareStrategies(rows, gt)
          : await ctx.session.compareStrategies(rows);
      return { recommended: out.recommended, strategies: out.strategies };
    },
  },
  {
    id: "suggest_pprl",
    description:
      "Inspect column names for sensitive PII and recommend privacy-preserving record linkage.",
    inputSchema: ROWS_OR_PATH_SCHEMA,
    handler: async (args, ctx): Promise<SkillResult> => {
      const rows = await resolveTable(args, ctx);
      const profile = profileForAgent(rows);
      const decision = selectStrategy(profile);
      return {
        has_sensitive: profile.has_sensitive,
        strategy: decision.strategy,
        recommendation: profile.has_sensitive
          ? "Sensitive fields detected; use privacy-preserving record linkage (PPRL)."
          : "No sensitive fields detected; PPRL is available but not required.",
      };
    },
  },
  {
    id: "agent_explain_pair",
    description: "Natural language explanation for a record pair.",
    inputSchema: EXPLAIN_PAIR_SCHEMA,
    handler: async (args): Promise<SkillResult> => {
      const rowA = args.record_a as Row | undefined;
      const rowB = args.record_b as Row | undefined;
      if (
        rowA === undefined ||
        rowB === undefined ||
        rowA === null ||
        rowB === null ||
        typeof rowA !== "object" ||
        typeof rowB !== "object"
      ) {
        return { error: "record_a and record_b are required objects" };
      }
      const mk = buildExplainMatchkey(rowA, rowB, args.fuzzy, args.exact);
      const explanation = explainPair(rowA, rowB, mk);
      return {
        score: explanation.score,
        confidence: explanation.confidence,
        explanation: explanation.explanation,
        field_scores: explanation.fieldScores,
        reasoning: explanation.reasoning,
      };
    },
  },
  {
    id: "agent_explain_cluster",
    description: "Explain why records are in the same cluster.",
    inputSchema: EXPLAIN_CLUSTER_SCHEMA,
    handler: async (args): Promise<SkillResult> => {
      const clusterId = args.cluster_id;
      return {
        cluster_id: clusterId !== undefined ? clusterId : null,
        note:
          "agent_explain_cluster requires a prior agent_deduplicate call. " +
          "Each dispatch is stateless; run agent_deduplicate first, then " +
          "inspect the clusters directly.",
      };
    },
  },
  {
    id: "controller_telemetry",
    description:
      "Return the AutoConfigController telemetry. Stateless dispatch cannot " +
      "read a prior run's telemetry; call auto_configure or agent_deduplicate " +
      "in the same invocation, which already returns telemetry inline.",
    inputSchema: CONTROLLER_TELEMETRY_SCHEMA,
    handler: async (): Promise<SkillResult> => {
      return {
        available: false,
        note:
          "controller_telemetry is per-session, but skill dispatch is " +
          "stateless. Call auto_configure or agent_deduplicate to get " +
          "telemetry alongside the result.",
      };
    },
  },
  {
    id: "agent_review_queue",
    description: "Get borderline pairs awaiting approval.",
    inputSchema: ROWS_OR_PATH_SCHEMA,
    handler: async (args, ctx): Promise<SkillResult> => {
      const rows = await resolveTable(args, ctx);
      const result = await dedupe(rows);
      const gated = gatePairs(result.scoredPairs);
      const pending = gated.needsReview.map((item) => ({
        id_a: item.idA,
        id_b: item.idB,
        score: item.score,
      }));
      return { pending, count: pending.length };
    },
  },
  {
    id: "agent_approve_reject",
    description: "Approve or reject a review queue pair.",
    inputSchema: APPROVE_REJECT_SCHEMA,
    handler: async (args, ctx): Promise<SkillResult> => {
      const decision = args.decision;
      if (decision !== "approve" && decision !== "reject") {
        return {
          error: `Invalid decision: ${String(decision)}. Use 'approve' or 'reject'.`,
        };
      }

      // Durable Learning Memory write, faithful to Python's
      // `_write_agent_correction`: source='agent', trust=0.5, empty
      // field/record hashes (staleness detection degrades gracefully),
      // original_score 0.0. The pair is canonicalized to (min, max) before
      // storage (the project-wide invariant, same as `add_correction`). When
      // no store is wired (edge path) the decision is still returned, matching
      // Python's `memory_store=None` branch.
      const idA = toInt(args.id_a);
      const idB = toInt(args.id_b);
      if (
        ctx.openMemoryStore !== undefined &&
        idA !== null &&
        idB !== null
      ) {
        const ca = Math.min(idA, idB);
        const cb = Math.max(idA, idB);
        const reason =
          typeof args.reason === "string" && args.reason.length > 0
            ? args.reason
            : null;
        const dataset =
          typeof args.dataset === "string" && args.dataset.length > 0
            ? args.dataset
            : ctx.dataset !== undefined
              ? ctx.dataset
              : null;
        const correction: Correction = {
          id: crypto.randomUUID(),
          idA: ca,
          idB: cb,
          decision,
          source: "agent",
          trust: 0.5,
          fieldHash: "",
          recordHash: "",
          originalScore: 0.0,
          matchkeyName: null,
          reason,
          dataset,
          createdAt: new Date(),
        };
        const store = await ctx.openMemoryStore();
        try {
          await store.addCorrection(correction);
        } finally {
          await store.close?.();
        }
      }

      // Response shape mirrors Python's `agent_approve_reject` handler:
      // {status, decision, job_name?, id_a, id_b, decided_by?}.
      const jobName =
        typeof args.job_name === "string" ? args.job_name : undefined;
      const decidedBy =
        typeof args.decided_by === "string" ? args.decided_by : undefined;
      return {
        status: "ok",
        decision,
        ...(jobName !== undefined ? { job_name: jobName } : {}),
        ...(args.id_a !== undefined ? { id_a: args.id_a } : {}),
        ...(args.id_b !== undefined ? { id_b: args.id_b } : {}),
        ...(decidedBy !== undefined ? { decided_by: decidedBy } : {}),
      };
    },
  },
  {
    id: "scan_quality",
    description:
      "Run GoldenCheck data quality scan on a dataset. Returns issues found " +
      "(encoding errors, Unicode problems, format violations) without applying " +
      "fixes. Requires goldencheck.",
    inputSchema: SCAN_QUALITY_SCHEMA,
    handler: async (args, ctx): Promise<SkillResult> => {
      try {
        await import("goldencheck" as string);
      } catch {
        return { error: "goldencheck not installed" };
      }
      // goldencheck present: still need the rows to scan.
      const rows = await resolveTable(args, ctx);
      return {
        error:
          "goldencheck integration is not wired in the TS core yet; " +
          `scanned 0 of ${rows.length} rows`,
      };
    },
  },
  {
    id: "fix_quality",
    description:
      "Run GoldenCheck scan and apply fixes to a dataset. Returns the fixed " +
      "data summary and a manifest of all fixes applied. Requires goldencheck.",
    inputSchema: FIX_QUALITY_SCHEMA,
    handler: async (args, ctx): Promise<SkillResult> => {
      try {
        await import("goldencheck" as string);
      } catch {
        return { error: "goldencheck not installed" };
      }
      const rows = await resolveTable(args, ctx);
      return {
        error:
          "goldencheck integration is not wired in the TS core yet; " +
          `fixed 0 of ${rows.length} rows`,
      };
    },
  },
  {
    id: "run_transforms",
    description:
      "Run GoldenFlow data transforms on a dataset. Normalizes phone numbers " +
      "(E.164), dates (ISO), categorical spelling, and Unicode issues. Returns " +
      "a manifest of transforms applied. Requires goldenflow.",
    inputSchema: ROWS_OR_PATH_SCHEMA,
    handler: async (args, ctx): Promise<SkillResult> => {
      try {
        await import("goldenflow" as string);
      } catch {
        return { error: "goldenflow not installed" };
      }
      const rows = await resolveTable(args, ctx);
      return {
        error:
          "goldenflow integration is not wired in the TS core yet; " +
          `transformed 0 of ${rows.length} rows`,
      };
    },
  },
  {
    id: "review_config",
    description:
      "Review a dedupe config and return verified config-improvement " +
      "suggestions (the healer). Opt-in: returns [] unless the suggest-wasm " +
      "backend is enabled.",
    inputSchema: REVIEW_CONFIG_SCHEMA,
    handler: async (args, ctx): Promise<SkillResult> => {
      const rows = await resolveTable(args, ctx);
      const explicit = args.config;
      const config: GoldenMatchConfig =
        explicit !== null && typeof explicit === "object" && !Array.isArray(explicit)
          ? (explicit as GoldenMatchConfig)
          : autoConfigureRows(rows);
      const verify = args.verify !== false; // default true (explicit request)
      const suggestions = await reviewConfig(rows, config, { verify });
      return {
        suggestions: serializeSuggestions(suggestions, { verified: verify }),
        count: suggestions.length,
      };
    },
  },
];

// ---------------------------------------------------------------------------
// Dispatcher
// ---------------------------------------------------------------------------

const SKILLS_BY_ID: ReadonlyMap<string, SkillDef> = new Map(
  AGENT_SKILLS.map((s) => [s.id, s]),
);

/**
 * Run an agent skill by id. The handler resolves its table from inline rows or
 * the injected `loadTable`. On any throw — unknown id, missing loader, handler
 * error — returns `{ error: <message> }` (never throws).
 */
export async function dispatchSkill(
  id: string,
  args: Record<string, unknown>,
  ctx: SkillContext,
): Promise<SkillResult> {
  const def = SKILLS_BY_ID.get(id);
  if (def === undefined) {
    return { error: `Unknown skill: ${id}` };
  }
  try {
    return await def.handler(args, ctx);
  } catch (err) {
    return { error: String(err instanceof Error ? err.message : err) };
  }
}
